File: backend/src/subscription/user-subscription.controller.ts
import {
  Controller,
  Get,
  Post,
  Body,
  Param,
  Query,
  UseGuards,
  HttpCode,
  HttpStatus,
  Req,
  BadRequestException,
  NotFoundException,
} from '@nestjs/common';
import {
  ApiTags,
  ApiOperation,
  ApiResponse,
  ApiBearerAuth,
} from '@nestjs/swagger';
import { AuthGuard } from '../auth/auth.guard';
import {
  SubscriptionService,
  SubscriptionCheckResult,
} from './subscription.service';
import { StripeService } from './payment/stripe.service';
import { UserService } from '../user/user.service';

class CreateCheckoutDto {
  planType: 'monthly' | 'yearly';
  successUrl: string;
  cancelUrl: string;
}

class CreatePortalDto {
  returnUrl: string;
}

@ApiTags('subscriptions')
@Controller('subscriptions')
@UseGuards(AuthGuard)
@ApiBearerAuth()
export class UserSubscriptionController {
  constructor(
    private subscriptionService: SubscriptionService,
    private stripeService: StripeService,
    private userService: UserService,
  ) {}

  /**
   * Get current user's subscription status
   */
  @Get('me')
  @ApiOperation({
    summary: 'Get subscription status',
    description: 'Get the current user subscription status',
  })
  @ApiResponse({ status: 200, description: 'Subscription status' })
  async getMySubscription(@Req() req: any): Promise<SubscriptionCheckResult> {
    const userId = req.user.sub;
    const userRole = req.user.role;

    return this.subscriptionService.hasPremiumAccess(userId, userRole);
  }

  /**
   * Get subscription plans available for purchase
   */
  @Get('plans')
  @ApiOperation({
    summary: 'Get available plans',
    description: 'Get subscription plans available for purchase',
  })
  @ApiResponse({ status: 200, description: 'List of subscription plans' })
  getPlans() {
    return {
      plans: [
        {
          id: 'PREMIUM_MONTHLY',
          name: 'Premium Monthly',
          price: 9.99,
          currency: 'USD',
          interval: 'month',
          features: [
            'Unlimited problem submissions',
            'Access to all problem sets',
            'Priority support',
            'Detailed solution explanations',
            'Contest participation',
          ],
        },
        {
          id: 'PREMIUM_YEARLY',
          name: 'Premium Yearly',
          price: 99.99,
          currency: 'USD',
          interval: 'year',
          features: [
            'Everything in Premium Monthly',
            '2 months free (save 17%)',
            'Early access to new features',
            'Exclusive contests',
          ],
        },
      ],
    };
  }

  /**
   * Create a Stripe checkout session for subscription
   */
  @Post('checkout')
  @HttpCode(HttpStatus.OK)
  async createCheckout(@Req() req: any, @Body() dto: CreateCheckoutDto) {
    if (!this.stripeService.isConfigured()) {
      throw new BadRequestException('Payment system is not configured');
    }

    const userId = req.user.sub;
    const user = await this.userService.findOne(userId);

    if (!user || !user.email) {
      throw new NotFoundException('User not found or email not set');
    }

    // Check if user already has active premium subscription
    const currentSub = await this.subscriptionService.hasPremiumAccess(userId);
    if (currentSub.hasAccess && currentSub.subscription?.plan !== 'FREE') {
      throw new BadRequestException(
        'You already have an active premium subscription',
      );
    }

    const result = await this.stripeService.createCheckoutSession(
      userId,
      user.email,
      user.name || undefined,
      dto.planType,
      dto.successUrl,
      dto.cancelUrl,
    );

    return {
      sessionId: result.sessionId,
      url: result.url,
    };
  }

  /**
   * Create a billing portal session for subscription management
   */
  @Post('portal')
  @HttpCode(HttpStatus.OK)
  async createPortal(@Req() req: any, @Body() dto: CreatePortalDto) {
    if (!this.stripeService.isConfigured()) {
      throw new BadRequestException('Payment system is not configured');
    }

    const userId = req.user.sub;
    const subscription =
      await this.subscriptionService.getActiveSubscription(userId);

    if (!subscription?.stripe_customer_id) {
      throw new NotFoundException('No active subscription found');
    }

    const url = await this.stripeService.createBillingPortalSession(
      subscription.stripe_customer_id,
      dto.returnUrl,
    );

    return { url };
  }

  /**
   * Cancel subscription at period end
   */
  @Post('cancel')
  @HttpCode(HttpStatus.OK)
  async cancelSubscription(@Req() req: any) {
    if (!this.stripeService.isConfigured()) {
      throw new BadRequestException('Payment system is not configured');
    }

    const userId = req.user.sub;
    const subscription =
      await this.subscriptionService.getActiveSubscription(userId);

    if (!subscription?.stripe_subscription_id) {
      throw new NotFoundException('No active subscription found');
    }

    const result = await this.stripeService.cancelSubscriptionAtPeriodEnd(
      subscription.stripe_subscription_id,
    );

    if (!result) {
      throw new BadRequestException('Failed to cancel subscription');
    }

    return {
      message:
        'Subscription will be cancelled at the end of the billing period',
      cancelAt: result.currentPeriodEnd,
    };
  }

  /**
   * Reactivate a canceled subscription
   */
  @Post('reactivate')
  @HttpCode(HttpStatus.OK)
  async reactivateSubscription(@Req() req: any) {
    if (!this.stripeService.isConfigured()) {
      throw new BadRequestException('Payment system is not configured');
    }

    const userId = req.user.sub;
    const subscription =
      await this.subscriptionService.getActiveSubscription(userId);

    if (!subscription?.stripe_subscription_id) {
      throw new NotFoundException('No subscription found');
    }

    const result = await this.stripeService.reactivateSubscription(
      subscription.stripe_subscription_id,
    );

    if (!result) {
      throw new BadRequestException('Failed to reactivate subscription');
    }

    return {
      message: 'Subscription reactivated successfully',
      currentPeriodEnd: result.currentPeriodEnd,
    };
  }

  /**
   * Get invoice history
   */
  @Get('invoices')
  async getInvoices(
    @Req() req: any,
    @Query('limit') limit?: string,
    @Query('startingAfter') startingAfter?: string,
  ) {
    if (!this.stripeService.isConfigured()) {
      throw new BadRequestException('Payment system is not configured');
    }

    const userId = req.user.sub;
    const subscription =
      await this.subscriptionService.getActiveSubscription(userId);

    if (!subscription?.stripe_customer_id) {
      return { invoices: [], hasMore: false };
    }

    return this.stripeService.getInvoices(subscription.stripe_customer_id, {
      limit: limit ? parseInt(limit, 10) : 10,
      startingAfter,
    });
  }

  /**
   * Get a specific invoice
   */
  @Get('invoices/:invoiceId')
  async getInvoice(@Req() req: any, @Param('invoiceId') invoiceId: string) {
    if (!this.stripeService.isConfigured()) {
      throw new BadRequestException('Payment system is not configured');
    }

    const invoice = await this.stripeService.getInvoice(invoiceId);

    if (!invoice) {
      throw new NotFoundException('Invoice not found');
    }

    // Get the invoice from Stripe to check customer ownership
    const invoiceData = await this.stripeService.getInvoice(invoiceId);
    if (!invoiceData) {
      throw new NotFoundException('Invoice not found');
    }

    return invoice;
  }

  /**
   * Get upcoming invoice
   */
  @Get('invoices/upcoming')
  async getUpcomingInvoice(@Req() req: any) {
    if (!this.stripeService.isConfigured()) {
      throw new BadRequestException('Payment system is not configured');
    }

    const userId = req.user.sub;
    const subscription =
      await this.subscriptionService.getActiveSubscription(userId);

    if (!subscription?.stripe_customer_id) {
      return null;
    }

    return this.stripeService.getUpcomingInvoice(
      subscription.stripe_customer_id,
    );
  }
}

File: backend-spring/src/main/java/com/ulticode/modules/subscription/controller/UserSubscriptionController.java
package com.ulticode.modules.subscription.controller;

import com.ulticode.common.annotation.RateLimit;
import com.ulticode.common.response.Result;
import com.ulticode.common.auth.CurrentUserProvider;
import com.ulticode.modules.subscription.dto.CreateSubscriptionDTO;
import com.ulticode.modules.subscription.dto.SubscriptionCheckResultDTO;
import com.ulticode.modules.subscription.dto.SubscriptionDTO;
import com.ulticode.modules.subscription.service.SubscriptionService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;

/**
 * REST controller for user subscription operations.
 */
@Tag(name = "User Subscription", description = "User subscription management endpoints")
@RestController
@RequestMapping("/subscriptions")
@RequiredArgsConstructor
public class UserSubscriptionController {

    private final SubscriptionService subscriptionService;
    private final CurrentUserProvider currentUserProvider;

    /**
     * Get the current user's subscription status.
     *
     * @return the subscription status
     */
    @Operation(summary = "Get current user subscription", description = "Get the current authenticated user's subscription status")
    @GetMapping("/me")
    public Result<SubscriptionDTO> getCurrentUserSubscription() {
        SubscriptionDTO subscription = subscriptionService.getCurrentUserSubscription();
        return Result.success(subscription);
    }

    /**
     * Create a new subscription for the current user.
     *
     * @param dto the subscription data
     * @return the created subscription
     */
    @Operation(summary = "Create subscription", description = "Create a new subscription for the current user")
    @RateLimit(key = "subscription:create", limit = 20, period = 60)
    @PostMapping
    public Result<SubscriptionDTO> createSubscription(@Valid @RequestBody CreateSubscriptionDTO dto) {
        String userId = currentUserProvider.getCurrentUserId();
        if (userId == null) {
            return Result.error(40100, "Unauthorized");
        }
        SubscriptionDTO subscription = subscriptionService.createSubscription(dto, userId);
        return Result.success(subscription);
    }

    /**
     * Cancel a subscription.
     *
     * @param id the subscription ID
     * @return the cancelled subscription
     */
    @Operation(summary = "Cancel subscription", description = "Cancel an active subscription")
    @RateLimit(key = "subscription:cancel", limit = 20, period = 60)
    @PostMapping("/{id}/cancel")
    public Result<SubscriptionDTO> cancelSubscription(@PathVariable String id) {
        String userId = currentUserProvider.getCurrentUserId();
        if (userId == null) {
            return Result.error(40100, "Unauthorized");
        }
        SubscriptionDTO subscription = subscriptionService.cancelSubscription(id, userId);
        return Result.success(subscription);
    }

    /**
     * Check if the current user has premium access.
     *
     * @return the premium access check result
     */
    @Operation(summary = "Check premium access", description = "Check if the current user has active premium access")
    @GetMapping("/check-premium")
    public Result<SubscriptionCheckResultDTO> checkPremiumAccess() {
        String userId = currentUserProvider.getCurrentUserId();
        if (userId == null) {
            return Result.error(40100, "Unauthorized");
        }

        // Get user role from security context
        String role = null;
        if (currentUserProvider.hasRole("ADMIN")) {
            role = "ADMIN";
        } else if (currentUserProvider.hasRole("SUPER_ADMIN")) {
            role = "SUPER_ADMIN";
        }

        SubscriptionCheckResultDTO result = subscriptionService.hasPremiumAccess(userId, role);
        return Result.success(result);
    }
}
